guide.md

```diff
@@ -0,0 +1,324 @@
+from Bazel, allowing it to provision external dependencies via the Nix package
+manager. Nix provides strong hermeticity and reproducibility guarantees for all
+build products, which are defined in a language also called Nix. A sizeable
+collection of such package definitions is available in the Nixpkgs repository,
+which can then be used freely from Bazel.
+
+# A basic Bazel project
+
+Let's start with a simple example, on the Bazel "happy path": building a simple
+C/C++ project. The project structure we'll work with is available [here] if
+you'd prefer to not go through the steps yourself.
+
+We start with an ordinary Bazel workspace setup:
+
+```
+- src
+  - BUILD
+  - hello-world.cc
+- WORKSPACE
+```
+
+where WORKSPACE is empty, and BUILD contains a single `cc_binary` rule:
+
+```bzl
+cc_binary(
+    name = "hello-world",
+    srcs = ["hello-world.cc"],
+)
+```
+
+and `src/hello-world.cc` is simply
+
+```cpp
+#include <iostream>
+
+int main(int argc, char** argv) {
+  std::cout << "Hello world!" << std::endl;
+  return 0;
+}
+```
+
+Verify that this builds:
+
+```
+> bazel run //src:hello-world
+INFO: Analyzed target //src:hello-world (0 packages loaded, 0 targets configured).
+INFO: Found 1 target...
+Target //src:hello-world up-to-date:
+  bazel-bin/src/hello-world
+INFO: Elapsed time: 0.063s, Critical Path: 0.00s
+INFO: 1 process: 1 internal.
+INFO: Build completed successfully, 1 total action
+INFO: Build completed successfully, 1 total action
+Hello world!
+```
+
+# The lack of hermeticity
+
+Bazel is valued for its hermetic and reproducible builds, but this example
+already shows that that isn't completely accurate: we haven't told Bazel what
+C++ compiler to use or where to find it and all the auxiliary tools it relies
+on (linker, assembler, and the like). It simply looks for appropriately-named
+executables in the PATH and sets up a C/C++ compiler toolchain called
+`@local_config_cc` that "should work". 
+
+This is problematic, and not just because of academic concerns regarding the
+precise meaning of "hermeticity": build products output by different versions
+of the same compiler can vary greatly, let alone between different compilers.
+Bazel, by default, chooses GCC on a Linux machine and Clang/LLVM on a macOS
+machine, which can mean that the same bit of hot code, say, vectorizes
+correctly on one platform but not on the other due to some difference between
+the two compilers, resulting in wildly different performance characteristics.
+
+There are Bazel-native ways to mitigate these issues, but they fall short of true
+hermeticity: for instance, `--incompatible_strict_action_env` runs actions in
+an environment that does not inherit a `PATH` from the shell environment, but
+instead [sets it to][path-link] the static value `/bin:/usr/bin:/usr/local/bin`.
+This means that as long as your compiler tools are in one of those paths, which
+they almost certainly are on \*nix systems, nothing changes.
```

Comment:
Generally true, but for the builtin C/C++ toolchain example this is irrelevant. The default C/C++ toolchain setup happens in a repository rule which is unaffected by `--incompatible_strict_action_env`.

Reply:
Took this out.